extension/chrome/elements/compose-modules/compose-recipients-module.ts

```diff
@@ -798,58 +807,71 @@ export class ComposeRecipientsModule extends ViewModule<ComposeView> {
         '<img src="/img/svgs/close-icon-black.svg" class="close-icon-black svg remove-reciepient">');
       $(el).find('.action_retry_pubkey_fetch').click(this.view.setHandler(async () => await this.refreshRecipients(), this.view.errModule.handle('refresh recipient')));
       $(el).find('.remove-reciepient').click(this.view.setHandler(element => this.removeRecipient(element.parentElement!), this.view.errModule.handle('remove recipient')));
-    } else if (contact === PUBKEY_LOOKUP_RESULT_WRONG) {
+    } else if (sortedPubKeyInfos === PUBKEY_LOOKUP_RESULT_WRONG) {
       recipient.status = RecipientStatus.WRONG;
       this.view.errModule.debug(`renderPubkeyResult: Setting email to wrong / misspelled in harsh mode: ${recipient.email}`);
       $(el).attr('title', 'This email address looks misspelled. Please try again.');
       $(el).addClass("wrong");
-    } else if (contact.pubkey && ((contact.expiresOn || Infinity) <= Date.now() || contact.pubkey.usableForEncryptionButExpired)) {
-      recipient.status = RecipientStatus.EXPIRED;
-      $(el).addClass("expired");
-      Xss.sanitizePrepend(el, '<img src="/img/svgs/expired-timer.svg" class="revoked-or-expired">');
-      $(el).attr('title', 'Does use encryption but their public key is expired. You should ask them to send ' +
-        'you an updated public key.' + this.recipientKeyIdText(contact));
-    } else if (contact.revoked) {
-      recipient.status = RecipientStatus.REVOKED;
-      $(el).addClass("revoked");
-      Xss.sanitizePrepend(el, '<img src="/img/svgs/revoked.svg" class="revoked-or-expired">');
-      $(el).attr('title', 'Does use encryption but their public key is revoked. You should ask them to send ' +
-        'you an updated public key.' + this.recipientKeyIdText(contact));
-    } else if (contact.pubkey) {
-      recipient.status = RecipientStatus.HAS_PGP;
-      $(el).addClass('has_pgp');
-      Xss.sanitizePrepend(el, '<img class="lock-icon" src="/img/svgs/locked-icon.svg" />');
-      $(el).attr('title', 'Does use encryption' + this.recipientKeyIdText(contact));
+    } else if (sortedPubKeyInfos.length) {
+      // New logic
+      // - if there is at least one valid (non-expired, non-revoked) public key, then it's HAS_PGP
+      // - else if there is at least one expired public key, then it's EXPIRED
+      // - else if there is at least one revoked key, then REVOKED
+      // - else it's NO_PGP
+      const firstKeyInfo = sortedPubKeyInfos[0];
+      if (!firstKeyInfo.revoked && !PubKeyInfoUtil.isExpired(firstKeyInfo)) {
+        recipient.status = RecipientStatus.HAS_PGP;
+        $(el).addClass('has_pgp');
+        Xss.sanitizePrepend(el, '<img class="lock-icon" src="/img/svgs/locked-icon.svg" />');
+        $(el).attr('title', 'Does use encryption\n' + this.publicKeysToRenderedText(sortedPubKeyInfos));
+      } else if (PubKeyInfoUtil.isExpired(firstKeyInfo)) {
```

Comment:
The code will get here if the key is both revoked and expired, right?
So the key will be marked as "EXPIRED" and potentially usable.
While revoked keys should not be used under any circumstances as they were compromised.
So you need to move the `revoked` clause up one level. Only remaining non-revoked keys can be "EXPIRED"
There might be (in rare cases) keys left that are neither usableForEncryption, nor revoked, nor expired (example -- weak keys). So it makes sense to check `KeyUtil.expired(firstKeyInfo.pubkey)` in the one-before-last `else if` clause to label the key as "EXPIRED", all other keys can be classified as NO_PGP. Or, maybe, @tomholub has an idea on how to mark weak keys?

Reply:
For now I will change order of checks, so check for revoked will be first.